### PR TITLE
Improve TileMap performances by using quadrants only for rendering

### LINKED ISSUE #79658: PathFollower2D Does NOT Update the Position and Collision of Child Tilemap with collision_animatable True
### Godot version

4.1.x

### System information

Windows 11

### Issue description

When a Tilemap node is attached as a child node of PathFollower2D node, with `collision_animatable` on in the inspector, the tilemap will stuck at (0, 0) if you scripted the movement of the PathFollower2D node with `progress` changed; while the tiles run as usual if `collision_animatable` is **false**.
This is very weird because the documentation said that "you can make Animatablebody2D move with script", and I used but it feedbacked a strange result, which is unacceptable and unwilling to see.
More detailed info please see the minimal project

P.S. Only under PathFollower2D, you add a RemoteTransformer2D that linked to a Tilemap can the tiles move as usual with `collision_animatable` on.

### Steps to reproduce

In the minimal reproduction project:

1. Make the `collision_animatable` of Tilemap and run the project, and you can see the tilemap moves as usual
2. Then tick on the property and run the game again, in which you can see the tilemap stuck at (0, 0) permanently.

### Minimal reproduction project

[bug report.zip](https://github.com/godotengine/godot/files/12094949/bug.report.zip)

### PR DESCRIPTION
**Pull request by groud**
A bit like #74603, but adds a ton of additional optimization and cleanup. 

All TileMap update are now pushed at the end of the frame, though you may trigger this update earlier on by calling `force_update()`. Also, I added a `notify_runtime_tile_data_update` so that you can tell the TileMap that some runtime data was updated, without triggering a direct update.

The TileMap now uses a `dirty` structure (with flags and a list of modified cells) to know what was modified since the last update.  Each TileMap's subsystem (rendering, physics, scenes...) then uses those flags and list of dirty cells to smartly update only what is needed. Unlike before, cells are only grouped in quadrants where needed. That means that updating a single cell does not update hundreds of others (at least on the physics matter)  for nothing.

This improvement, added to a reduction of the number of allocations, lead to a x2.8 performance improvement on TileMap updates. Using this small benchmark script, I measured FPSs:
```
extends Node2D

func _process(delta):
	random_tile_map_update()
	
func random_tile_map_update():
	for x in range(100):
		for y in range(100):
			if $TileMap.get_cell_source_id(0, Vector2i(x,y)) == -1:
				$TileMap.set_cell(0, Vector2i(x,y), 0, Vector2i(randi_range(0, 2), randi_range(0, 2)), 0)
			elif randi_range(0, 100) == 0:
				$TileMap.set_cell(0, Vector2i(x,y), 0, Vector2i(randi_range(0, 2), randi_range(0, 2)), 0)
```

Right now this is a draft as I need to fix CI, add compatibility methods for `set_quadrant_size` and document a few things. But it's ready for testing otherwise. :)  

Probably fixes: https://github.com/godotengine/godot/issues/79658

### CODE REVIEW

File: scene/2d/tile_map.compat.inc [hunk 1]
```diff
@@ -34,8 +34,24 @@ Rect2i TileMap::_get_used_rect_bind_compat_78328() {
...
+void TileMap::_force_update_compat_81070(int p_layer) {
+	notify_runtime_tile_data_update();
+	force_update();
+}
+
 void TileMap::_bind_compatibility_methods() {
 	ClassDB::bind_compatibility_method(D_METHOD("get_used_rect"), &TileMap::_get_used_rect_bind_compat_78328);
+	ClassDB::bind_compatibility_method(D_METHOD("set_quadrant_size", "quadrant_size"), &TileMap::_set_quadrant_size_compat_81070);
+	ClassDB::bind_compatibility_method(D_METHOD("get_quadrant_size"), &TileMap::_get_quadrant_size_compat_81070);
+	ClassDB::bind_compatibility_method(D_METHOD("force_update", "layer"), &TileMap::_force_update_compat_81070);
```
Comment (dsnopek) [inline scene/2d/tile_map.compat.inc:54]: For the CI error about the hash to go away, this needs the `DEFVAL(-1)` that was registered with the method previously

**Comment by groud** [comment]
> What happened to the layer argument in force_update()? It doesn't exist in notify_runtime_tile_data_update().

Ah good point. I removed it as update are ran once per frame by the TileMap node itself, and not by each layer individually. But it likely makes sense to add it to `notify_runtime_tile_data_update`. Will do that.

I'll have a look to the bugs.

**Comment by groud** [comment]
Done. I fixed the bugs and added an optional argument to `notify_runtime_tile_data_update`, let me know if you still post any bug.

**Comment by KoBeWi** [comment]
The previous problems are fixed, but I'm getting this error:
```
ERROR: No tiles for the given body RID 74178380178984.
   at: (C:\godot_source\scene/2d/tile_map.cpp:3392)
```
Might be caused by me deleting all collision polygons on a layer. Not sure if intended.

File: scene/2d/tile_map.compat.inc [hunk 2]
```diff
@@ -34,8 +34,24 @@ Rect2i TileMap::_get_used_rect_bind_compat_78328() {
...
+void TileMap::_set_quadrant_size_compat_81070(int p_quadrant_size) {
+	set_rendering_quadrant_size(p_quadrant_size);
+}
+
+int TileMap::_get_quadrant_size_compat_81070() const {
+	return get_rendering_quadrant_size();
+}
+
+void TileMap::_force_update_compat_81070(int p_layer) {
+	notify_runtime_tile_data_update();
```
Comment (dsnopek) [inline scene/2d/tile_map.compat.inc:46]: `notify_runtime_tile_data_update()` takes a layer argument. Should `p_layer` be passed along here?

**Comment by groud** [comment]
> The previous problems are fixed, but I'm getting this error:

This is the message given by `get_coords_for_body_rid` when it cannot find the provided RID. I think this message could maybe be removed, as you might call it with an RID you are not sure is part of a TileMap. 

I am not sure what you mean by "deleting all collision polygons on a layer", I guess by using a runtime update ? If that's the case, it should not prevent `get_coords_for_body_rid` to work correctly (in case you are colliding with a TileMap), so there might be a bug.

File: scene/2d/tile_map.compat.inc (hunk 2 above)
Comment (groud) [reply scene/2d/tile_map.compat.inc:46]: Ah, my bad, that's an oversight.

**Comment by dsnopek** [comment]
As discussed on RocketChat, the compatibility methods added here will provide compatibility for GDExtension, however, renaming these methods will still affect GDScript and C# compatibility. I don't how popular/obscure these methods are, so maybe it's OK if GDScript/C# code calling the old names breaks, but that's a discussion that should be had!

If we want to preserve GDScript/C# compatibility, the old methods could be marked as deprecated and simply call the new methods. (And, if we went that way, the compatibility methods for GDExtension could be removed.)

**Comment by KoBeWi** [comment]
> I guess by using a runtime update ?

Yes, I use it do disable collisions on a layer.

File: scene/2d/tile_map.h [hunk 3]
```diff
@@ -382,15 +462,20 @@ class TileMap : public Node2D {
 	virtual Rect2 _edit_get_rect() const override;
 #endif
 
+#ifndef DISABLE_DEPRECATED
+	// Deprecated methods
```
Comment (KoBeWi) [inline scene/2d/tile_map.h:466]:
```suggestion
	// Deprecated methods.
```
The comment is unnecessary tbh.

File: scene/2d/tile_map.h [hunk 4]
```diff
@@ -310,15 +393,15 @@ class TileMapLayer : public RefCounted {
...
-	void force_update();
-
 	// Fixing and clearing methods.
 	void fix_invalid_tiles();
 
 	// Find coords for body.
 	bool has_body_rid(RID p_physics_body) const;
 	Vector2i get_coords_for_body_rid(RID p_physics_body) const; // For finding tiles from collision.
+
+	// Destructor
```
Comment (AThousandShips) [inline scene/2d/tile_map.h:403]: Unnecessary

File: scene/2d/tile_map.h [hunk 5]
```diff
@@ -89,75 +89,134 @@ class TerrainConstraint {
...
+
+public:
+	Vector2i quadrant_coords;
+	SelfList<CellData>::List cells;
+	RID canvas_item;
+
+	SelfList<DebugQuadrant> dirty_quadrant_list_element;
+
+	// For those, copy everything but SelfList elements.
+	DebugQuadrant(const DebugQuadrant &c) :
```
Comment (AThousandShips) [inline scene/2d/tile_map.h:163]:
```suggestion
	DebugQuadrant(const DebugQuadrant &p_c) :
```

File: scene/2d/tile_map.h [hunk 6]
```diff
@@ -89,75 +89,134 @@ class TerrainConstraint {
...
+		cell = c.cell;
+		occluders = c.occluders;
+		bodies = c.bodies;
+		navigation_regions = c.navigation_regions;
+		scene = c.scene;
+		runtime_tile_data_cache = c.runtime_tile_data_cache;
 	}
-
-	TileMapQuadrant(const TileMapQuadrant &q) :
+	CellData(const CellData &c) :
```
Comment (AThousandShips) [inline scene/2d/tile_map.h:132]:
```suggestion
	CellData(const CellData &p_c) :
```

File: scene/2d/tile_map.h [hunk 7]
```diff
@@ -89,75 +89,134 @@ class TerrainConstraint {
...
-	// This is kind of workaround, it would be better to sort the cells directly in the "cells" set instead.
-	RBMap<Vector2i, Vector2> map_to_local;
-	RBMap<Vector2, Vector2i, CoordsWorldComparator> local_to_map;
-
-	// Debug.
-	RID debug_canvas_item;
+	// Rendering.
+	SelfList<CellData> debug_quadrant_list_element;
 
 	// Rendering.
```
Comment (KoBeWi) [inline scene/2d/tile_map.h:102]:
```suggestion
```
Repeated.

File: scene/2d/tile_map.h [hunk 8]
```diff
@@ -89,75 +89,134 @@ class TerrainConstraint {
...
-		occluders = q.occluders;
-		bodies = q.bodies;
-		navigation_regions = q.navigation_regions;
+	TileData *runtime_tile_data_cache = nullptr;
+
+	// List elements.
+	SelfList<CellData> dirty_list_element;
+
+	// For those, copy everything but SelfList elements.
+	void operator=(const CellData &c) {
```
Comment (AThousandShips) [inline scene/2d/tile_map.h:123]:
```suggestion
	void operator=(const CellData &p_c) {
```

File: scene/2d/tile_map.h [hunk 9]
```diff
@@ -89,75 +89,134 @@ class TerrainConstraint {
...
+	};
+
+	Vector2i quadrant_coords;
+	SelfList<CellData>::List cells;
+	List<RID> canvas_items;
+
+	SelfList<RenderingQuadrant> dirty_quadrant_list_element;
+
+	// For those, copy everything but SelfList elements.
+	RenderingQuadrant(const RenderingQuadrant &c) :
```
Comment (AThousandShips) [inline scene/2d/tile_map.h:201]:
```suggestion
	RenderingQuadrant(const RenderingQuadrant &p_c) :
```

File: scene/2d/tile_map.h [hunk 10]
```diff
@@ -89,75 +89,134 @@ class TerrainConstraint {
...
+		cell = p_other.cell;
+		occluders = p_other.occluders;
+		bodies = p_other.bodies;
+		navigation_regions = p_other.navigation_regions;
+		scene = p_other.scene;
+		runtime_tile_data_cache = p_other.runtime_tile_data_cache;
 	}
-
-	TileMapQuadrant(const TileMapQuadrant &q) :
+	CellData(const CellData &p_other) :
```
Comment (KoBeWi) [inline scene/2d/tile_map.h:132]:
```suggestion

	CellData(const CellData &p_other) :
```
Missing blank line.

File: scene/2d/tile_map.h [hunk 11]
```diff
@@ -89,75 +89,134 @@ class TerrainConstraint {
...
+			debug_quadrant_list_element(this),
+			rendering_quadrant_list_element(this),
 			dirty_list_element(this) {
 	}
 };
 
+struct DebugQuadrant : public RefCounted {
+	GDCLASS(DebugQuadrant, RefCounted);
+
+public:
```
Comment (KoBeWi) [inline scene/2d/tile_map.h:155]: This is a `struct`, so it's already public.
Not sure why didn't you use `class` though.

File: scene/2d/tile_map.h [hunk 12]
```diff
@@ -166,6 +225,34 @@ class TileMapLayer : public RefCounted {
...
+		DIRTY_FLAGS_TILE_MAP_TEXTURE_FILTER,
+		DIRTY_FLAGS_TILE_MAP_TEXTURE_REPEAT,
+		DIRTY_FLAGS_TILE_MAP_TILE_SET,
+		DIRTY_FLAGS_TILE_MAP_QUADRANT_SIZE,
+		DIRTY_FLAGS_TILE_MAP_COLLISION_ANIMATABLE,
+		DIRTY_FLAGS_TILE_MAP_COLLISION_VISIBILITY_MODE,
+		DIRTY_FLAGS_TILE_MAP_NAVIGATION_VISIBILITY_MODE,
+		DIRTY_FLAGS_TILE_MAP_Y_SORT_ENABLED,
+		DIRTY_FLAGS_TILE_MAP_RUNTIME_UPDATE,
+		DIRTY_FLAGS_MAX,
```
Comment (KoBeWi) [inline scene/2d/tile_map.h:253]: Are that many flags actually needed? Some of them seem unused or have duplicate meaning.

Also you are using them with a bool array, but the assigned value resets in the middle (`DIRTY_FLAGS_TILE_MAP_IN_TREE `). This doesn't seem right.

File: scene/2d/tile_map.cpp [hunk 13]
```diff
@@ -40,175 +40,291 @@
 #include "servers/navigation_server_3d.h"
 #endif // DEBUG_ENABLED
 
-Vector2i TileMapLayer::_coords_to_quadrant_coords(const Vector2i &p_coords) const {
-	int quad_size = get_effective_quadrant_size();
+#define TILE_MAP_DEBUG_QUADRANT_SIZE 16
```
Comment (KoBeWi) [inline scene/2d/tile_map.cpp:43]:
```suggestion
constexpr int TILE_MAP_DEBUG_QUADRANT_SIZE = 16;
```
`constexpr` is preferred over macros when defining constants (see canvas item editor plugin for example).
Also you can get rid of the weird assign below.

File: scene/2d/tile_map.cpp [hunk 14]
```diff
@@ -40,175 +40,291 @@
...
-	rs->canvas_item_set_z_index(q.debug_canvas_item, RS::CANVAS_ITEM_Z_MAX - 1);
-	rs->canvas_item_set_parent(q.debug_canvas_item, tile_map_node->get_canvas_item());
+	if (forced_cleanup) {
+		for (KeyValue<Vector2i, Ref<DebugQuadrant>> &kv : debug_quadrant_map) {
+			// Free the quadrant.
+			Ref<DebugQuadrant> &debug_quadrant = kv.value;
+			if (debug_quadrant->canvas_item.is_valid()) {
+				rs->free(debug_quadrant->canvas_item);
+			}
+			debug_quadrant_map.erase(debug_quadrant->quadrant_coords);
```
Comment (KoBeWi) [inline scene/2d/tile_map.cpp:67]: Wouldn't it be better/enough to do `debug_quadrant_map.clear()` after the loop?

File: scene/2d/tile_map.cpp [hunk 15]
```diff
@@ -40,175 +40,291 @@
...
-			p_coords.x > 0 ? p_coords.x / quad_size : (p_coords.x - (quad_size - 1)) / quad_size,
-			p_coords.y > 0 ? p_coords.y / quad_size : (p_coords.y - (quad_size - 1)) / quad_size);
+			p_coords.x > 0 ? p_coords.x / debug_quadrant_size : (p_coords.x - (debug_quadrant_size - 1)) / debug_quadrant_size,
+			p_coords.y > 0 ? p_coords.y / debug_quadrant_size : (p_coords.y - (debug_quadrant_size - 1)) / debug_quadrant_size);
 }
 
-HashMap<Vector2i, TileMapQuadrant>::Iterator TileMapLayer::_create_quadrant(const Vector2i &p_qk) {
-	TileMapQuadrant q;
-	q.coords = p_qk;
+void TileMapLayer::_debug_update() {
```
Comment (KoBeWi) [inline scene/2d/tile_map.cpp:53]: If it's debug then it should be all wrapped in `#ifdef DEBUG_ENABLED` I think. Only parts of the debug update are conditional...

File: scene/2d/tile_map.cpp [hunk 16]
```diff
@@ -243,84 +358,213 @@ void TileMapLayer::_rendering_update_dirty_quadrants(SelfList<TileMapQuadrant>::
...
-		q_list_element = q_list_element->next();
-	}
+		dirty_rendering_quadrant_list.clear();
 
-	// Reset the drawing indices.
-	if (_rendering_quadrant_order_dirty) {
-		int index = -(int64_t)0x80000000; //always must be drawn below children.
+		// Reset the drawing indices.
+		{
+			int index = -(int64_t)0x80000000; //always must be drawn below children.
```
Comment (KoBeWi) [inline scene/2d/tile_map.cpp:394]:
```suggestion
			int index = -(int64_t)0x80000000; // Always must be drawn below children.
```

File: scene/2d/tile_map.cpp [hunk 17]
```diff
@@ -373,60 +612,127 @@ void TileMapLayer::_rendering_draw_quadrant_debug(TileMapQuadrant *p_quadrant) {
...
+	const Ref<TileSet> &tile_set = tile_map_node->get_tileset();
+	Transform2D gl_transform = tile_map_node->get_global_transform();
+	RID space = tile_map_node->get_world_2d()->get_space();
+	PhysicsServer2D *ps = PhysicsServer2D::get_singleton();
 
-		// Recreate bodies and shapes.
-		for (const Vector2i &E_cell : q.cells) {
-			TileMapCell c = get_cell(E_cell, true);
+	// Recreate bodies and shapes.
+	TileMapCell c = r_cell_data.cell;
```
Comment (KoBeWi) [inline scene/2d/tile_map.cpp:686]: The TileMapCells are small enough that it doesn't matter if you pass them by reference or by value, but seems like you are doing it inconsistently.

File: scene/2d/tile_map.cpp [hunk 18]
```diff
@@ -3172,23 +3294,11 @@ void TileMap::set_y_sort_enabled(bool p_enable) {
...
-void TileMap::_recreate_internals() {
 	for (Ref<TileMapLayer> &layer : layers) {
-		layer->recreate_internals();
+		layer->notify_tile_map_change(TileMapLayer::DIRTY_FLAGS_TILE_MAP_Y_SORT_ENABLED);
 	}
+	emit_signal(CoreStringNames::get_singleton()->changed);
+	update_configuration_warnings();
 }
 
 /////////////////////////////// Rendering //////////////////////////////////////
```
Comment (KoBeWi) [inline scene/2d/tile_map.cpp:3304]: Probably unrelated to this PR, but the methods below don't look like rendering related.

**Comment by KoBeWi** [comment]
I did a more thorough test and something bad is happening when the TileMap is removed and re-added to scene tree. I'm removing tile collisions with `tile_data.set_collision_polygons_count(0, 0)`. It works fine, but after the TileMap is re-added (I didn't test exactly) the collisions are back.

File: scene/2d/tile_map.cpp (hunk 15 above)
Comment (groud) [reply scene/2d/tile_map.cpp:53]: I don't know. From what I could read, it seems like it is not done for drawing collision shape in general. So, I am not sure we can add a condition here.

File: scene/2d/tile_map.h (hunk 12 above)
Comment (groud) [reply scene/2d/tile_map.h:253]: Some flags seem unused as they are not, like, used specifically by their name. However they are checked in the `_debug_update` by a loop:
```
	for (int i = 0; i < DIRTY_FLAGS_MAX; i++) {
		if (dirty.flags[i]) {
			anything_changed = true;
			break;
		}
	}
```
I guess I could replace several flags by a single "something changed" flag, but TBH, I prefer to keep it that way in case we need more optimization. It's a bit weird as it means we have a lot of flags, but at least it is consistent and helps when implementing optimizations.

**Comment by groud** [comment]
> I did a more thorough test and something bad is happening when the TileMap is removed and re-added to scene tree. I'm removing tile collisions with tile_data.set_collision_polygons_count(0, 0). It works fine, but after the TileMap is re-added (I didn't test exactly) the collisions are back.

That was my bad, it should be solved now.

File: scene/2d/tile_map.cpp (hunk 15 above)
Comment (KoBeWi) [reply scene/2d/tile_map.cpp:52]: Collision shapes can only draw in debug build.
What I mean is that the debug code should not run at all in release builds. Meantime the `_debug_update()` is always called and wastes performance when not needed.

Comment (groud) [reply scene/2d/tile_map.cpp:52]: > Collision shapes can only draw in debug build.

If that's the case, then yes, I can wrap the code in `#ifdef DEBUG_ENABLED` then. I'll do that.

**Review by KoBeWi — APPROVED** [review]

File: doc/classes/TileMap.xml [hunk 19]
```diff
@@ -73,13 +76,11 @@
...
 		</method>
-		<method name="force_update">
+		<method name="force_update" is_deprecated="true">
 			<return type="void" />
 			<param index="0" name="layer" type="int" default="-1" />
 			<description>
-				Triggers an update of the TileMap. If [param layer] is provided and is positive, only updates the given layer.
-				[b]Note:[/b] The TileMap node updates automatically when one of its properties is modified. A manual update is only needed if runtime modifications (implemented in [method _tile_data_runtime_update]) need to be applied.
-				[b]Warning:[/b] Updating the TileMap is computationally expensive and may impact performance. Try to limit the number of updates and the tiles they impact (by placing frequently updated tiles in a dedicated layer for example).
+				This method is deprecated. See [method notify_runtime_tile_data_update] and [method update_now].
```
Comment (YuriSizov) [inline doc/classes/TileMap.xml:83]:
```suggestion
				[i]Deprecated.[/i] See [method notify_runtime_tile_data_update] and [method update_now].
```

For consistency with other documentation.

File: scene/2d/tile_map.cpp [hunk 20]
```diff
@@ -40,175 +40,291 @@
...
 		}
-		q.occluders.clear();
 
-		// Those allow to group cell per material or z-index.
-		Ref<Material> prev_material;
-		int prev_z_index = 0;
-		RID prev_ci;
+		// Update all dirty quadrants.
+		for (SelfList<RenderingQuadrant> *quadrant_list_element = dirty_rendering_quadrant_list.first(); quadrant_list_element;) {
+			SelfList<RenderingQuadrant> *next_quadrant_list_element = quadrant_list_element->next(); // "Hack" to clear the list while iterating.
```
Comment (YuriSizov) [inline scene/2d/tile_map.cpp:275]: Would it be significantly slower to collect items to delete and then do the clearing? This feels very error prone.

**Review by YuriSizov — APPROVED** [review]
Skimmed through code because it's a lot, but aside from the for-loop hack nothing stood out. Everything seems well organized and well documented with comments.

I do have one question, though. Wouldn't it be more compatibility-friendly to keep `force_update` instead of introducing `update_now`? I understand that it is supposed to be used in places where `notify_runtime_tile_data_update` is now used. But since the logic needs to be changed anyway in projects, I feel like we can accept the changes in the method's behavior. Having two methods triggering an immediate update, according to their name, is pretty confusing, even if one is deprecated. Plus `force_update` seems more appropriate and consistent for the public API and our codebase in general.

File: scene/2d/tile_map.cpp (hunk 20 above)
Comment (YuriSizov) [reply scene/2d/tile_map.cpp:275]: Well my question is if it's used for a good reason or because it was just simpler to write code this way 🙃 There is no natural increment in the loop's definition, so a tiny mistake can make it go on forever. Same applies to any other place that uses this trick.

Comment (groud) [reply scene/2d/tile_map.cpp:275]: The problem is that the whole point of SelfList is to limit allocations to a minimum. If I allocate a structure to list the objects to delete, there's little gain into using a SelfList at all. 

I believe (didn't test) the performance gain likely depends on how many quadrants you have, but I guess it could get quite big with Y-sorting enabled, as the quadrant size is reduced to 1 tile = 1 quadrant.

In any case, I called that a "hack" as, like, it's a bit of a weird implementation. But TBH, I think it's perfectly fine. There's no reason the pointer should point to garbage, even if the SelfList implementation changed, as we are basically pointing an object that is allocated by the TileMap. I could remove the comment if that's a bit bothering.

**Comment by groud** [comment]
> I do have one question, though. Wouldn't it be more compatibility-friendly to keep force_update instead of introducing update_now? I understand that it is supposed to be used in places where notify_runtime_tile_data_update is now used. But since the logic needs to be changed anyway in projects, I feel like we can accept the changes in the method's behavior. Having two methods triggering an immediate update, according to their name, is pretty confusing, even if one is deprecated. Plus force_update seems more appropriate and consistent for the public API and our codebase in general.

We discussed that a bit. Before this PR, `force_update` used to do two things: 1) notify _runtime_tile_data* methods might lead to different results and 2) trigger a direct update instead of waiting till the end of the frame.

Initially, I made `force_update` only do 2), and created `notify_runtime_tile_data_update` to cover 1). This is a cleanup and a bit of a optimization, as notifying several times would not trigger several unneeded updates. The naming was likely the most expected according to what each of those new function do. 

However, the main issue is that `force_update()` seemed mostly used for 2). This means that, in most cases, this would break users' code without them knowing what's wrong or triggering any error.

**Comment by YuriSizov** [comment]
> This means that, in most cases, this would break users' code without them knowing what's wrong or triggering any error.

Right, but using this method as-is is not optimal, correct? So we would want users to update to using `notify_runtime_tile_data_update` regardless. And the breakage would only come in the form of things not updating if `update_now/force_update` is called without `notify_runtime_tile_data_update` prior to it? I think that would be okay.

**Comment by groud** [comment]
> Right, but using this method as-is is not optimal, correct? So we would want users to update to using notify_runtime_tile_data_update regardless.

Depends, but yes. Most of the time it's more optimal to only use `notify_runtime_tile_data_update`

> And the breakage would only come in the form of things not updating if update_now/force_update is called without notify_runtime_tile_data_update prior to it? I think that would be okay.

The problem is that "things not updating" could be subtle. It might be some collision shape or custom data not updating, which might be hard to spot.

> Edit: To clarify, we would mention this change in the migration article.

We can do that. I guess it's up to the production team to decide what you folks prefer.

To be clear, I think `force_update` is a better name, I am just worried that the upgrade path is likely going to be a bit confusing for some people. That why we went for the `update_now` rename initially, but if it's well documented, that should be fine I guess.

**Comment by akien-mga** [comment]
Will need a rebase after #80354 which split the `4.0.stable.expected` file into one for `4.0 -> 4.1` and one for `4.1 -> master`.

File: scene/2d/tile_map.cpp [hunk 21]
```diff
@@ -2759,114 +2888,115 @@ int TileMap::get_selected_layer() const {
...
+				set_notify_local_transform(true);
+			}
+		} break;
 	}
 }
 
-void TileMap::queue_update_dirty_quadrants() {
-	if (pending_update || !is_inside_tree()) {
+#ifndef DISABLE_DEPRECATED
+// Deprecated methods
```
Comment (AThousandShips) [inline scene/2d/tile_map.cpp:2972]:
```suggestion
// Deprecated methods.
```

**Comment by groud** [comment]
Put that PR to draft for now, `collision_animatable` isn't working anymore and I need to fix it :)

File: misc/extension_api_validation/4.1-stable.expected [hunk 22]
```diff
@@ -116,3 +116,11 @@ Validate extension JSON: API was removed: classes/GLTFDocumentExtensionPhysics
...
 
 Excluded unexposed classes from extension_api.json.
+
+GH-81070
+--------
+Validate extension JSON: API was removed: classes/TileMap/methods/get_quadrant_size
+Validate extension JSON: API was removed: classes/TileMap/methods/set_quadrant_size
+Validate extension JSON: API was removed: classes/TileMap/properties/cell_quadrant_size
+
+cell_quadrant_size/quadrant_size of the TileMap API was renamed to rendering_quadrant_size.
```
Comment (AThousandShips) [inline misc/extension_api_validation/4.1-stable.expected:126]:
```suggestion
cell_quadrant_size/quadrant_size of the TileMap API was renamed to rendering_quadrant_size.

```
Missing trailing newline

**Comment by groud** [comment]
> Put that PR to draft for now, collision_animatable isn't working anymore and I need to fix it :)

This is now fixed.

**Comment by YuriSizov** [comment]
Regarding `force_update`/`update_now`. Suppose the deprecation of the old method is reasonable (as the old name suggests it will force an update, and the new behavior won't do that, it acts more like flushing a queue). Can we maybe pick a different name for `update_now` to make it less colliding with the old `force_update`?

@akien-mga suggested `apply_queued_update`, which sounds fine to me. Maybe some variation of that.

**Comment by groud** [comment]
> @akien-mga suggested apply_queued_update (or just apply_update), which sounds fine to me. Maybe some variation of that.

Yeah, that could work fine too. I thought about `preemptive_update` maybe, or `early_update`, as it makes it clear that the TileMap would update anyway. `apply_queued_update` would work, though maybe a bit confusing as users do not manually "queue an update" themselves.

I don't know, I can't decide ahah.

**Comment by YuriSizov** [comment]
They kind of do with the notify method, no? But in that case, just `apply_update` should be a good option? As long as the docs make it clear that updates are "applied" automatically and this method is does not need to be called normally.

**Comment by groud** [comment]
> They kind of do with the notify method, no? 

If they use the runtime TileData update (which I believe not a lot of users need) they might. 

But the current `update_now` function is a bit unrelated. Basically all TileMap update (like, physics, rendering, navigation, scenes, etc...) are pushed to the end of the frame. This makes it so all TileMap internals (bodies, canvas items, child scenes, etc...) are not accessible right after a property change on the TileMap. 

For example, if you paint a scene tile, the actual scene will be added only at the end of the frame. That's where you might want to call `update_now` right away, so that you could access some internal data earlier on:
```gdscript
$TileMap.set_cell(0, position, 0, Vector2i(0,0)) # Set a scene tile.
$TileMap.get_child_count() # returns 0

$TileMap.update_now()
$TileMap.get_child_count() # returns 1
```
Basically, that means that the main use case for `update_now` does not need any call the other notification function.

File: scene/2d/tile_map.cpp [hunk 23]
```diff
@@ -2759,114 +2906,114 @@ int TileMap::get_selected_layer() const {
...
+}
+#endif
+
+void TileMap::queue_internal_update() {
+	if (pending_update) {
 		return;
 	}
 	pending_update = true;
-	call_deferred(SNAME("_update_dirty_quadrants"));
+	call_deferred(SNAME("_internal_update"));
```
Comment (YuriSizov) [inline scene/2d/tile_map.cpp:3005]:
```suggestion
	callable_mp(this, &TileMap::_internal_update).call_deferred();
```

And you don't need to bind it anymore.

**Comment by YuriSizov** [comment]
> Basically, that means that the main use case for `update_now` does not need any call the other notification function.

Okay, I'm a bit confused about how old code maps to the new one. Can't it be just `update` then? What `now` brings to the table? It seems to just force an update.

**Comment by groud** [comment]
> > Basically, that means that the main use case for `update_now` does not need any call the other notification function.
> 
> Okay, I'm a bit confused about how old code maps to the new one. Can't it be just `update` then? What `now` brings to the table? It seems to just force an update.

Well, yeah. That would work perfectly, lol. 

I think I did not consider the idea because it was not possible before due to `CanvasItem.update()`, but since it was renamed in https://github.com/godotengine/godot/pull/64377, it should be perfectly doable now. A simple `update()` would work.

**Comment by groud** [comment]
Renamed to `update_internals`, as discussed on the maintainers' chat.

**Review by YuriSizov — APPROVED** [review]
I'll just re-approve to confirm this consensus :)

**Comment by akien-mga** [comment]
Thanks!
